### PR TITLE
TDL-17610: Multiple apps support and data fetching

### PR DESCRIPTION
**Pull request by karanpanchal-crest**
# Description of change
 - Allows user to configure multiple app_ids to extract data
 - Fetches data for all app_ids if not provided (the current version only fetches data for default app_id which cannot be changed)

# Manual QA steps
 - Verify that if app_ids are not available in config then we get all apps data and if app_ids are given then we get those apps data
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by dbshah1212 — APPROVED** [review]

File: tests/unittests/test_app_ids_configurable.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        """
+            To verify that if app_ids is blank string or empty string in config than select all apps.
+        """
+        config = {"app_ids": " "}
+        stream_obj = Features(config)
+        self.assertEqual(stream_obj.app_ids,  "expandAppIds(\"*\")", "Both values are not same")
+    
+    def test_app_ids_comaseperated_string_in_config(self):
+        """
+            To verify that f app_ids is comma seperated string in config than get list of those app_ids.
```
Comment (savan-chovatiya) [inline tests/unittests/test_app_ids_configurable.py:24]:
```suggestion
            To verify that if app_ids is comma seperated string in config then get list of those app_ids.
```

File: tests/unittests/test_app_ids_configurable.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        """
+            To verify that if app_ids is not in config than select all apps.
+        """
+        config = {}
+        stream_obj = Features(config)
+        self.assertEqual(stream_obj.app_ids,  "expandAppIds(\"*\")", "Both values are not same")
+        
+    def test_app_ids_empty_in_config(self):
+        """
+            To verify that if app_ids is blank string or empty string in config than select all apps.
```
Comment (savan-chovatiya) [inline tests/unittests/test_app_ids_configurable.py:16]:
```suggestion
            To verify that if app_ids is blank string or empty string in config then select all apps.
```

File: tests/unittests/test_app_ids_configurable.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import unittest
+from tap_pendo.streams import Features
+
+
+class TestAppIdConfiguration(unittest.TestCase):
+    def test_app_ids_not_in_config(self):
+        """
+            To verify that if app_ids is not in config than select all apps.
```
Comment (savan-chovatiya) [inline tests/unittests/test_app_ids_configurable.py:8]:
```suggestion
            To verify that if app_ids is not in config then select all apps.
```

**Review by savan-chovatiya — APPROVED** [review]

File: tests/unittests/test_app_ids_configurable.py (hunk 1 above)
Comment (karanpanchal-crest) [reply tests/unittests/test_app_ids_configurable.py]: Added suggested changes.

File: tests/unittests/test_app_ids_configurable.py (hunk 2 above)
Comment (karanpanchal-crest) [reply tests/unittests/test_app_ids_configurable.py]: Added suggested changes.

File: tests/unittests/test_app_ids_configurable.py (hunk 3 above)
Comment (karanpanchal-crest) [reply tests/unittests/test_app_ids_configurable.py]: Added suggested changes.

File: tests/unittests/test_app_ids_configurable.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class TestAppIdConfiguration(unittest.TestCase):
+    @mock.patch("tap_pendo.utils.parse_args", side_effect=lambda required_config_keys: Namespace(config={"start_date": "", "x_pendo_integration_key": "", "period":""},  discover=True))
+    @mock.patch("tap_pendo.discover_streams", side_effect=lambda config: {})
+    def test_app_ids_not_in_config(self, mocked_discover_stream, mocked_parse_args):
+        """
+            To verify that if app_ids is not in configure file then select all apps and run discover mode.
+        """
+        main()
```
Comment (savan-chovatiya) [inline tests/unittests/test_app_ids_configurable.py:14]: Add at least one assertion about app_ids value in config or any other relevant assertion.

File: tap_pendo/__init__.py [hunk 5]
```diff
@@ -151,14 +155,32 @@ def sync(config, state, catalog):
...
+    invalid_app_ids = []
+    if app_ids != "expandAppIds(\"*\")":
+        app_ids = app_ids.split(",")
+        invalid_app_ids = list(filter(isInteger, app_ids))
+
+    args.config["app_ids"] = app_ids
+
     if args.discover:
-        do_discover(args.config)
+        do_discover(args.config, invalid_app_ids)
```
Comment (dbshah1212) [inline tap_pendo/__init__.py:183]: Raise exception for the invalid_app_ids here only. Either move the whole code in do_discover or put an exception here only.

Comment (karanpanchal-crest) [reply tap_pendo/__init__.py]: Raised exception for the invalid_app_ids inside the  main function

File: tap_pendo/__init__.py [hunk 6]
```diff
@@ -151,13 +151,34 @@ def sync(config, state, catalog):
...
+    # If app_ids is given in configure file then check all app_ids are numeric or not
+    invalid_app_ids = []
+    if app_ids != "expandAppIds(\"*\")":
+        app_ids = app_ids.split(",")
+        invalid_app_ids = list(filter(isInteger, app_ids))
+
+    args.config["app_ids"] = app_ids
+
     if args.discover:
+        # If non numeric app_ids are given in configure file then raise exception
```
Comment (dbshah1212) [inline tap_pendo/__init__.py:179]: put it before the discovery if, there is no harm to checking this in sync mode.

Comment (karanpanchal-crest) [reply tap_pendo/__init__.py]: Put the exception before "if args.discover" condition

**Review by dbshah1212 — APPROVED** [review]

File: tests/tap_tester/test_all_fields.py [hunk 7]
```diff
@@ -87,6 +87,12 @@ def test_run(self):
                     
                 # verify all fields for each stream are replicated
                 self.assertSetEqual(expected_all_keys, actual_all_keys)
+                
+                # Verify we have more than one app data
+                # below four streams are independent of the app_id
+                if stream not in ["accounts", "visitors", "metadata_accounts", "metadata_visitors"]:
+                    records_appid_set = set([message.get('data').get('app_id') for message in messages.get("messages")])
+                    self.assertGreater(len(records_appid_set), 1, msg=f"We have only one app's records for {stream}")
```
Comment (kspeer825) [inline tests/tap_tester/test_all_fields.py:95]: This is a good assertion to cover the case when the `app_ids` are not set in the config. Could you add a configurable properties test (at the tap-tester level) to verify the tap can execute a sync when called via the orchestrator with `app_ids` set in the config?

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/tap_tester/test_all_fields.py (hunk 7 above)
Comment (karanpanchal-crest) [reply tests/tap_tester/test_all_fields.py:95]: Thanks @kspeer825, I have added the integration test ["test_app_ids"](https://github.com/singer-io/tap-pendo/blob/TDL-17610-multiple-apps-support-and-data-fetching/tests/tap_tester/test_app_ids.py) as you suggested.

**Review by Vi6hal — APPROVED** [review]
reviewed and tested these changes

**Review by dsprayberry — APPROVED** [review]

File: tests/tap_tester/test_app_ids.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                fields_from_field_level_md)
+
+        self.run_and_verify_sync(conn_id)
+
+        synced_records = runner.get_records_from_target_output()
+
+        # Verify no unexpected streams were replicated
+        synced_stream_names = set(synced_records.keys())
+
+        # Skipping below streams due to zero records for given start date
```
Comment (bhtowles) [inline tests/tap_tester/test_app_ids.py:73]: What does this comment mean?  Where are streams being skipped due to zero records?

File: tests/tap_tester/test_app_ids.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+class PendoMultiAppIdsTest(TestPendoBase):
+    def name(self):
+        return "pendo_multi_app_ids_test"
+
+    def test_run(self):
+        expected_streams = {"features", "pages", "events", "guides", "track_types", "track_events"}
+        self.run_test(expected_streams=expected_streams, app_ids="-323232", is_multi_apps=False)
+        self.run_test(expected_streams=expected_streams, app_ids=None, is_multi_apps=True)
+        self.run_test(expected_streams={"visitors", "visitor_history"}, app_ids=None, is_multi_apps=True)
```
[the remark this answers is not in the document]
Comment (RushiT0122) [reply tests/tap_tester/test_app_ids.py:15]: Only "accounts", "visitors", "metadata_accounts" and "metadata_visitors" streams are not supported for this test and to ensure maximum test coverage within the three-hour maximum execution time in cci, we handle `visitor_history` stream in a specific manner. Note earlier we used to skip this stream due same limitation.

File: tests/tap_tester/test_app_ids.py (hunk 8 above)
Comment (RushiT0122) [reply tests/tap_tester/test_app_ids.py:73]: This should be removed

File: tests/tap_tester/test_app_ids.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                return return_value
+
+            return return_value
+        else:
+            return super().get_properties()
+
+    def run_test(self, expected_streams, app_ids, is_multi_apps, start_date=None):
+        """
+        - Verify tap syncs records for multiple app_ids if no app_ids provided
+        - Verify tap syncs records for specific app_id if
```
[the remark this answers is not in the document]
Comment (RushiT0122) [reply tests/tap_tester/test_app_ids.py]: Fixed

**Comment by RushiT0122** [comment]
@bhtowles please re-review.

File: tests/tap_tester/test_app_ids.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhtowles) [reply tests/tap_tester/test_app_ids.py:15]: Ok

**Review by bhtowles — APPROVED** [review]

**Pull request merged**